File: 17.py
from aoc import *

DAY = 17

s = get_input(DAY)
# s = get_example(DAY)

import collections
import math

ls = s.splitlines()

instructions = list(map(int, ls[4].split(": ")[1].split(",")))


def run_program(a):
    b = 0
    c = 0

    i = 0
    while i < len(instructions):
        opcode = instructions[i]
        operand = instructions[i + 1]

        combo_operand = operand
        if combo_operand == 4:
            combo_operand = a
        elif combo_operand == 5:
            combo_operand = b
        elif combo_operand == 6:
            combo_operand = c
        elif combo_operand == 7:
            assert False

        if opcode == 0:
            a >>= combo_operand
        elif opcode == 1:
            b ^= operand
        elif opcode == 2:
            b = combo_operand % 8
        elif opcode == 3:
            if a != 0:
                i = operand
                continue
        elif opcode == 4:
            b ^= c
        elif opcode == 5:
            yield combo_operand % 8
        elif opcode == 6:
            b = a >> combo_operand
        elif opcode == 7:
            c = a >> combo_operand
        else:
            assert False

        i += 2


def part1():
    a = int(ls[0].split(": ")[1])
    return ",".join(map(str, run_program(a)))


def part2(a=0, i=len(instructions)-1):
    if i < 0:
        return a

    current_a = a * 8
    for j in range(8):
        program_output = run_program(current_a + j)
        if all(instructions[i] == next(program_output) for i in range(i, len(instructions))):
            ans = part2(current_a + j, i - 1)
            if ans > 0:
                return ans

    return -1

# submit(DAY, part1(), 1)
# submit(DAY, part2(), 2)

File: kotlin/src/17.kt
import kotlin.time.measureTimedValue

fun main() {
    val registers = readInput()
    val instructions = readInput()[0].substringAfter(": ").split(',').map { it.toInt() }

    fun runProgram(baseA: Long) = sequence {
        var a = baseA
        var b = 0L
        var c = 0L

        var i = 0
        while (i < instructions.size) {
            val operand = instructions[i + 1].toLong()
            val comboOperand = when (operand) {
                4L -> a
                5L -> b
                6L -> c
                else -> operand
            }

            when(instructions[i]) {
                0 -> a = a shr comboOperand.toInt()
                1 -> b = b xor operand
                2 -> b = comboOperand % 8
                3 -> if (a != 0L) {
                    i = operand.toInt()
                    continue
                }
                4 -> b = b xor c
                5 -> yield((comboOperand % 8).toInt())
                6 -> b = a shr comboOperand.toInt()
                7 -> c = a shr comboOperand.toInt()
            }

            i += 2
        }
    }

    fun part1(): String {
        val a = registers[0].substringAfter(": ").toLong()
        return runProgram(a).joinToString(",")
    }

    fun part2(a: Long = 0L, i: Int = instructions.lastIndex): Long {
        if (i < 0)
            return a

        val currentA = a * 8
        for (j in 0 ..< 8) {
            val programOutput = runProgram(currentA + j).iterator()
            if ((i .. instructions.lastIndex).all { instructions[it] == programOutput.next() }) {
                val ans = part2(currentA + j, i - 1)
                if (ans > 0)
                    return ans
            }
        }

        return -1
    }

    println(measureTimedValue { part1() }.let { "${it.value} (${it.duration})" })
    println(measureTimedValue { part2() }.let { "${it.value} (${it.duration})" })
}
